src/providers/pdal/CMakeLists.txt

```
@@ -257,6 +257,12 @@ if (WITH_GUI)
   add_dependencies(provider_pdal_gui_a ui)
 endif()
 
+if(${CMAKE_SYSTEM_NAME} MATCHES "BSD")
+  set(CMAKE_THREAD_LIBS -pthread)
+  set(CMAKE_USE_PTHREADS ON)
+  set(CMAKE_EXE_LINKER_FLAGS -pthread)
+endif(${CMAKE_SYSTEM_NAME} MATCHES "BSD")
+
```

Would a more generic (and more modern cmake style) approach work as well?

```suggestion
set(THREADS_PREFER_PTHREAD_FLAG ON)
find_package(Threads REQUIRED)
target_link_libraries(provider_pdal_gui_a PRIVATE Threads::Threads)
```

---

Strange, this is actually already done. What is the error you are getting?

---

```
[6/10726] Linking CXX executable output/lib/qgis/pdal_wrench
FAILED: output/lib/qgis/pdal_wrench
: && ccache /usr/bin/c++ -Wall -Wextra -Wno-long-long -Wformat-security -Wno-strict-aliasing -Wnon-virtual-dtor -Wno-redundant-move -Wno-misleading-indentation -Wreturn-type-c-linkage -Woverloaded-virtual -Wimplicit-fallthrough -Qunused-arguments -Qunused-arguments src/providers/pdal/CMakeFiles/pdal_wrench.dir/pdal_wrench_autogen/mocs_compilation.cpp.o src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/main.cpp.o src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/alg.cpp.o src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/boundary.cpp.o src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/clip.cpp.o src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/density.cpp.o src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/info.cpp.o src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/merge.cpp.o src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/thin.cpp.o src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/to_raster.cpp.o src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/to_raster_tin.cpp.o src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/to_vector.cpp.o src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/translate.cpp.o src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/utils.cpp.o src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/vpc.cpp.o src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/tile/tile.cpp.o src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/tile/BufferCache.cpp.o src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/tile/Cell.cpp.o src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/tile/FileProcessor.cpp.o src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/tile/Las.cpp.o src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/tile/TileGrid.cpp.o src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/tile/ThreadPool.cpp.o src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/tile/Writer.cpp.o -o output/lib/qgis/pdal_wrench -L/root/QGIS/build/src/core   -L/root/QGIS/build/src/gui   -L/usr/local/lib -Wl,-rpath,/root/QGIS/build/src/core:/root/QGIS/build/src/gui:/usr/local/lib:  -lpdalcpp  /usr/local/lib/libpdal_util.so  /usr/local/lib/libgdal.so && :
ld: error: undefined symbol: pthread_create
>>> referenced by ThreadPool.cpp
>>>               src/providers/pdal/CMakeFiles/pdal_wrench.dir/__/__/__/external/pdal_wrench/tile/ThreadPool.cpp.o:(std::__1::__libcpp_thread_create(pthread**, void* (*)(void*), void*))
c++: error: linker command failed with exit code 1 (use -v to see invocation)
```

---

> Would a more generic (and more modern cmake style) approach work as well?

Nope,

`set(CMAKE_EXE_LINKER_FLAGS -pthread)` is mandatory

---

> `set(CMAKE_EXE_LINKER_FLAGS -pthread)` is mandatory

Do you have a reference to that?
If I understand correctly, this would mean that all downstream apps of a lib using pthread would need to add a platform specific condition, which sounds different to all the cmake efforts to transfer such information via targets.